### ISSUE TITLE
Happy Eyeballs

### ISSUE THREAD

**Issue by swankjesse** [square/okhttp#506]
http://en.wikipedia.org/wiki/Happy_Eyeballs

**Comment by sriharim25** [comment]
@dlubarov , @JakeWharton do you folks happen to be working on bringing Happy Eyeballs support to OkHttp? There are more DHCPv6-based IPv6 networks every day, especially in corporate settings, and the ability to pick the better transport in parallel will greatly benefit user experience.

**Comment by swankjesse** [comment]
It’s possible to implement this manually by making two HTTP requests yourself and picking the one that returns first. Doing it inside of OkHttp is more difficult. For example, what should the interceptors see? We might be able to do further work later to unblock this, but for now it seems very difficult to do well.

**Comment by mikelambert** [comment]
So this is causing issues for IPv6 http requests with React Native, linked above.

It sounds like the workaround suggested above is to do two requests in the above layer, with one forced to IPv4. Curious, how does one force IPv4 in an okhttp request?

Do I need to resolve the DNS to the IPv4 and IPv6 address and add in Host headers manually? Or is there some easier way?

**Comment by swankjesse** [comment]
Implement the Dns interface and strip the addresses you don't want. Be careful to not return an empty list; that case needs to throw an exception.

**Comment by ericmguimaraes** [comment]
Hey guys, by any chance you still planing on doing that? I'm facing the same problem here and the workaround doesn't really look good. It would be great to have a more sophisticated approach built-in the lib.

**Comment by yschimke** [comment]
Linking directly to RFC - https://tools.ietf.org/html/rfc8305

**Comment by medusar** [comment]
So, what is the plan? Will okhttp support Happy Eyeballs ?

**Comment by dbrgn** [comment]
In the [docs](https://square.github.io/okhttp/), I read:

> OkHttp perseveres when the network is troublesome: it will silently recover from common connection problems. If your service has multiple IP addresses OkHttp will attempt alternate addresses if the first connect fails. This is necessary for IPv4+IPv6 and services hosted in redundant data centers.

However, when I run this code in an API 21 (5.0) emulator...

```java
public class MainActivity extends AppCompatActivity {

    @SuppressLint("StaticFieldLeak")
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        new AsyncTask<Void, Void, Void>() {
            @Override
            protected Void doInBackground(Void... voids) {
                OkHttpClient client = new OkHttpClient();
                String url = "https://directory.spaceapi.io/";
                Request request = new Request.Builder()
                        .url(url)
                        .build();
                try (Response response = client.newCall(request).execute()) {
                    Log.i("MainActivity", "Body: " + response.body().string());
                } catch (IOException e) {
                    e.printStackTrace();
                }
                return null;
            }
        }.execute();
    }
}
```

...it fails with the following exception:

```
02-29 17:27:37.293 5855-5873/? W/System.err: java.net.ConnectException: Failed to connect to directory.spaceapi.io/2a01:4f8:1c0c:8127::2:443
02-29 17:27:37.293 5855-5873/? W/System.err:     at okhttp3.internal.connection.RealConnection.connectSocket(RealConnection.kt:285)
02-29 17:27:37.293 5855-5873/? W/System.err:     at okhttp3.internal.connection.RealConnection.connect(RealConnection.kt:195)
02-29 17:27:37.293 5855-5873/? W/System.err:     at okhttp3.internal.connection.ExchangeFinder.findConnection(ExchangeFinder.kt:233)
02-29 17:27:37.293 5855-5873/? W/System.err:     at okhttp3.internal.connection.ExchangeFinder.findHealthyConnection(ExchangeFinder.kt:107)
02-29 17:27:37.293 5855-5873/? W/System.err:     at okhttp3.internal.connection.ExchangeFinder.find(ExchangeFinder.kt:75)
02-29 17:27:37.293 5855-5873/? W/System.err:     at okhttp3.internal.connection.RealCall.initExchange$okhttp(RealCall.kt:245)
02-29 17:27:37.293 5855-5873/? W/System.err:     at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.kt:32)
02-29 17:27:37.293 5855-5873/? W/System.err:     at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:100)
02-29 17:27:37.293 5855-5873/? W/System.err:     at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.kt:82)
02-29 17:27:37.293 5855-5873/? W/System.err:     at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:100)
02-29 17:27:37.293 5855-5873/? W/System.err:     at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.kt:83)
02-29 17:27:37.293 5855-5873/? W/System.err:     at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:100)
02-29 17:27:37.293 5855-5873/? W/System.err:     at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.kt:74)
02-29 17:27:37.293 5855-5873/? W/System.err:     at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:100)
02-29 17:27:37.293 5855-5873/? W/System.err:     at okhttp3.internal.connection.RealCall.getResponseWithInterceptorChain$okhttp(RealCall.kt:197)
02-29 17:27:37.293 5855-5873/? W/System.err:     at okhttp3.internal.connection.RealCall.execute(RealCall.kt:148)
02-29 17:27:37.293 5855-5873/? W/System.err:     at ch.dbrgn.happyeyeballs.MainActivity$1.doInBackground(MainActivity.java:31)
02-29 17:27:37.293 5855-5873/? W/System.err:     at ch.dbrgn.happyeyeballs.MainActivity$1.doInBackground(MainActivity.java:23)
02-29 17:27:37.293 5855-5873/? W/System.err:     at android.os.AsyncTask$2.call(AsyncTask.java:288)
02-29 17:27:37.293 5855-5873/? W/System.err:     at java.util.concurrent.FutureTask.run(FutureTask.java:237)
02-29 17:27:37.293 5855-5873/? W/System.err:     at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:231)
02-29 17:27:37.293 5855-5873/? W/System.err:     at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112)
02-29 17:27:37.293 5855-5873/? W/System.err:     at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
02-29 17:27:37.293 5855-5873/? W/System.err:     at java.lang.Thread.run(Thread.java:818)
02-29 17:27:37.293 5855-5873/? W/System.err: Caused by: java.net.ConnectException: failed to connect to directory.spaceapi.io/2a01:4f8:1c0c:8127::2 (port 443) after 10000ms: isConnected failed: ENETUNREACH (Network is unreachable)
02-29 17:27:37.293 5855-5873/? W/System.err:     at libcore.io.IoBridge.isConnected(IoBridge.java:238)
02-29 17:27:37.293 5855-5873/? W/System.err:     at libcore.io.IoBridge.connectErrno(IoBridge.java:171)
02-29 17:27:37.294 5855-5873/? W/System.err:     at libcore.io.IoBridge.connect(IoBridge.java:122)
02-29 17:27:37.294 5855-5873/? W/System.err:     at java.net.PlainSocketImpl.connect(PlainSocketImpl.java:183)
02-29 17:27:37.294 5855-5873/? W/System.err:     at java.net.PlainSocketImpl.connect(PlainSocketImpl.java:456)
02-29 17:27:37.294 5855-5873/? W/System.err:     at java.net.Socket.connect(Socket.java:882)
02-29 17:27:37.294 5855-5873/? W/System.err:     at okhttp3.internal.platform.AndroidPlatform.connectSocket(AndroidPlatform.kt:57)
02-29 17:27:37.294 5855-5873/? W/System.err:     at okhttp3.internal.connection.RealConnection.connectSocket(RealConnection.kt:283)
02-29 17:27:37.294 5855-5873/? W/System.err: 	... 23 more
02-29 17:27:37.294 5855-5873/? W/System.err: Caused by: android.system.ErrnoException: isConnected failed: ENETUNREACH (Network is unreachable)
02-29 17:27:37.294 5855-5873/? W/System.err:     at libcore.io.IoBridge.isConnected(IoBridge.java:223)
02-29 17:27:37.294 5855-5873/? W/System.err: 	... 30 more
```

That host (directory.spaceapi.io) is accessible both via IPv4 and IPv6. However, the emulator only has link-local and site-local IPv6 addresses.

```
root@generic_x86_64:/ # ip addr
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host
       valid_lft forever preferred_lft forever
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1440 qdisc pfifo_fast state UP qlen 1000
    link/ether 52:54:00:12:34:56 brd ff:ff:ff:ff:ff:ff
    inet 10.0.2.15/24 brd 10.0.2.255 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 fec0::5054:ff:fe12:3456/64 scope site dynamic
       valid_lft 86236sec preferred_lft 14236sec
    inet6 fe80::5054:ff:fe12:3456/64 scope link
       valid_lft forever preferred_lft forever
3: sit0: <NOARP> mtu 1480 qdisc noop state DOWN
    link/sit 0.0.0.0 brd 0.0.0.0
```

Am I right when I assume that this connection failure is due to the missing happy eyeballs implementation?

**Comment by yschimke** [comment]
Does that host work from chrome on the same mobile?

**Comment by dbrgn** [comment]
> Does that host work from chrome on the same mobile?

Hmm, good question. From the built-in emulator browser I get a `net::ERR_SSL_VERSION_OR_CIPHER_MISMATCH` error. The target host has disabled TLS < 1.2. Could this be a TLS version mismatch masked as a IPv6 connection problem? The same code *does* work on an Android 9 emulator.

**Comment by yschimke** [comment]
I would suggest starting with a Stackoverflow question.  That request works for me with a JDK OkHttp request.  But I get an IPv4 address back.

But debugging on this github issue is likely to spam 8 people.

**Comment by yschimke** [comment]
One thing to consider is whether we should add suppressed errors when we fail after multiple hosts. 

If we fail with IPv4 with a useful error but swallow it to deliver the IPv6 error then it's misleading.

**Comment by taktv6** [comment]
Any idea on when we can have Happy Eyeballs with OkHTTP?

**Comment by swankjesse** [comment]
It's an invasive feature to add because there's extra complexity. If you need it, don't wait for us.

**Comment by hzhyan** [comment]
Why don't you implement the Happy Eyeballs algorithm？I think [dlubarov‘s submission](https://github.com/square/okhttp/commit/d037c8b339a693583c93d4bb9c94aa9e04f4ecd2) is a good way.
Removing the V6 address to solve this problem is really not a good solution,This is a step backwards. @yschimke

**Comment by yschimke** [comment]
@hzhyan I think you are misunderstanding my position

> Ideally the fix would be in OkHttp - Happy Eyeballs #506. But that isn't happening soon.
> 
> As a high touch solution you could disable IPv6 either on your device, or https://www.cactusvpn.com/tutorials/how-to-disable-ipv6-on-android/ or look into why your network dns or the remote service is advertising IPv6. It's generally a network misconfiguration somewhere if you are seeing IPv6 addresses from DNS that are getting blackholed.
> 
> The pragmatic solution is to implement a Dns decorator to filter out IPv6, but it's a bandaid https://github.com/yschimke/okurl/blob/e307022667e2beb474309af5b350cd241f2a9045/src/main/kotlin/com/baulsupp/okurl/network/DnsSelector.kt
> 
> But honestly it's good to know that OkHttp is doing the correct thing here and repairing the connection.

But we don't have it, and merging a patch from 2014 isn't trivial.  The good thing here for current users is that OkHttp recovers from the failing IPv6 results, the logging tools are enough to explain the behaviour and there are possible workarounds, but it's obviously not ideal.

**Comment by hzhyan** [comment]
> @hzhyan I think you are misunderstanding my position
> 
> > Ideally the fix would be in OkHttp - Happy Eyeballs #506. But that isn't happening soon.
> > As a high touch solution you could disable IPv6 either on your device, or https://www.cactusvpn.com/tutorials/how-to-disable-ipv6-on-android/ or look into why your network dns or the remote service is advertising IPv6. It's generally a network misconfiguration somewhere if you are seeing IPv6 addresses from DNS that are getting blackholed.
> > The pragmatic solution is to implement a Dns decorator to filter out IPv6, but it's a bandaid https://github.com/yschimke/okurl/blob/e307022667e2beb474309af5b350cd241f2a9045/src/main/kotlin/com/baulsupp/okurl/network/DnsSelector.kt
> > But honestly it's good to know that OkHttp is doing the correct thing here and repairing the connection.
> 
> But we don't have it, and merging a patch from 2014 isn't trivial. The good thing here for current users is that OkHttp recovers from the failing IPv6 results, the logging tools are enough to explain the behaviour and there are possible workarounds, but it's obviously not ideal.

Thanks, looking forward to your solution and upgrade.

**Comment by yschimke** [comment]
@hzhyan I'll quote @swankjesse on that "It's an invasive feature to add because there's extra complexity. If you need it, don't wait for us."

**Comment by hzhyan** [comment]
> @hzhyan I'll quote @swankjesse on that "It's an invasive feature to add because there's extra complexity. If you need it, don't wait for us."

OK

**Comment by sureshg** [comment]
https://download.java.net/java/early_access/loom/docs/api/java.base/java/util/concurrent/TaskSession.html#:~:text=API%20Note%3A-,the%20following%20is,-a%20more%20advanced - Seems like it will be easy to solve this in loom with the new structured concurrency API 😄

**Comment by andreialecu** [comment]
I don't mean to spam all participants to this thread, but I'm hoping that more datapoints are useful in order to help prioritize this issue.

If there's any way this would be taken another look at it would be amazing for the React Native community.

Random Android networking issues have been reported all over the ecosystem there over the past few years, and they seem to all be related to imperfect IPv6 setups. I left a more detailed writeup here and linked other issues: https://github.com/facebook/react-native/issues/32730#issuecomment-990764376

Also related: #6954 where there was a simple router glitch causing the problem, where restarting it fixed the problem, and not IPv6 misconfiguration. 

There's a workaround possible where prioritising IPv4 addresses also helps, detailed in: https://github.com/facebook/react-native/issues/29608#issuecomment-884521699

**Comment by SuperQ** [comment]
@swankjesse Thanks for getting this rolling.

I know there's concern about increasing the number of TCP sessions created.

One possible mitigation I was thinking about is this workflow. In the case of getting both an `AAAA` and `A` result. Delay connecting to the IPv4 address by some small number of milliseconds. This way we have a high chance of making the IPv6 connection before we even need to bother with the fallback mode.

Based on some simple assumptions:
* My observations so far show that IPv6 is working correctly for > 99.9% of users. Users affected by configured but broken IPv6  are a fairly small long-tail population.
* Somewhere around 40% of mobile users now have a IPv6 connection. (In my network stats, we see between 30 and 45% IPv6 to our CDN) 
* The vast majority of traffic is going to hit a CDN.
* The major CDNs have PoPs within 25ms of the majority of their users.

If we set a IPv4 fallback delay of 50ms, that should vastly reduce the number of dual connections, without majorly impacting user experience. Even a lower delay in the 20-30ms range should greatly reduce the extra connections without too much user visible impact.

**Comment by yschimke** [comment]
+1 I think a lot of this is discussed in https://datatracker.ietf.org/doc/html/rfc8305, hopefully that can give some best practice configurations.

OkHttp also have a route database so hopefully some of these are one offs per app launch.

**Comment by andreialecu** [comment]
Note that the issue isn't always about misconfigured IPv6 due to negligence. 

In one location, with a specific router, I've noticed that IPv6 randomly starts glitching at least once every month or so, until the router is reset or the connection is reconnected manually.

When it happens, it's pretty obvious, because a small subset of apps will start glitching and run into connection issues. However, most (high-profile) apps continue to work just fine.

It appears that node 17 made the network stack default to IPv6, which now exhibits the same problem.

I noticed that when the router IPv6 glitch occurs, connections to a MongoDB server from my dev machine take forever as well. It's mentioned here that NodeJS doesn't implement Happy Eyeballs (yet) either: https://www.reddit.com/r/ipv6/comments/qbr8jc/nodejs_finally_prefers_ipv6_addresses_over_ipv4/

**Comment by SuperQ** [comment]
@andreialecu Yes, there are a whole bunch of random ways users can have broken IPv6 (and IPv4) connectivity.

**Comment by treysis** [comment]
@andreialecu Not exactly. NodeJS doesn't prioritize IPv6 over IPv4. It just uses the addresses in the order returned by the OS. So in the default setting since NodeJS v17 it is up to the OS to prefer IPv6 over IPv4 or not.

Btw: The typical delay for Happy Eyeballs should be in the 200-300 ms range.

**Comment by andreialecu** [comment]
There's a write up of how this was implemented in OSX here:

https://mailarchive.ietf.org/arch/msg/v6ops/DYiI9v_O66RNbMJsx0NsatFkubQ/

> 	The updated implementation performs the following:
> 	- Query the DNS resolver for A and AAAA.
> 	   If the DNS records are not in the cache, the requests are sent back to back on the wire, AAAA first.
> 	- If the first reply we get is AAAA, we send out the v6 SYN immediately
> 	- If the first reply we get is A and we're expecting a AAAA, we start a 25ms timer
> 	   - If the timer fires, we send out the v4 SYN
> 	   - If we get the AAAA during that 25ms window, we move on to address selection
> 	- When we have a list of IP addresses (either from the DNS cache or by receiving them close together with v4 before v6),
> 	   we perform our own address selection algorithm to sort them. This algorithm uses historical RTT data to prefer addresses
> 	   that have lower latency - but has a 25ms leeway: if the historical RTT of two compared address are within 25ms of each
> 	   other, we use RFC3484 to pick the best one.
> 	- Once the list is sorted, we send out the SYN for the first address and start timers based on average and variance of the
> 	   historical TCP RTT. Roughly speaking, we start the second address around the same time we send out a SYN retransmission
> 	   for the first address.
> 	- The first address to reply with a SYN-ACK wins the race, we then cancel the other TCP connection attempts

.

**Comment by SuperQ** [comment]
My only caution is to use [RFC6724](https://datatracker.ietf.org/doc/html/rfc6724) instead of RFC3484. I've seen too many implementations (Linux libc6 was one of the worst) that numerically sort multiple results, leading to completely breaking any kind of DNS round-robin balancing.

**Comment by swankjesse** [comment]
Initial implementation is here: https://github.com/square/okhttp/pull/7035

In 5.0.0-alpha.4 you’ll be able to turn it on like this:

```
      OkHttpClient client = new OkHttpClient.Builder()
        .fastFallback(true)
        .build();
```

**Comment by swankjesse** [comment]
Next steps before 5.0.0-alpha.4:

 * [ ] Sort IPv6 and IPv4 addresses
 * [ ] Limit how many connections are attempted simultaneously
 * [ ] Unit tests for FastFallbackExchangeFinder using a fake clock

Next steps before closing this issue as complete:

 * [ ] Move the TLS handshake phase out of the race. As soon as we have a TCP handshake success we should declare the race as won, cancel other in-flight connects, and then perform the TLS handshake on the connection.
 * [ ] Follow guidance in @andreialecu’s comment on how it works in OSX.

In a separate issue we should investigate splitting IPv4 vs. IPv6 DNS lookups.

**Comment by swankjesse** [comment]
This is done.

**Comment by andreialecu** [comment]
Awesome, thank you so much for this!

I was wondering when a stable 5.0 release may be available. The milestone here doesn't help much (says past due by 1 year 🙂)
https://github.com/square/okhttp/milestone/21

**Comment by swankjesse** [comment]
If you're prepared to deal with small API changes between now and 5.0 final you can put it into production today. For OkHttp ‘alpha’ means ‘API instability’ but the code is extremely stable.

**Comment by balajithiru1** [comment]
Do we know when this feature will be out of alpha , so we can start using them in production?

**Comment by SuperQ** [comment]
It's marked "alpha", but according to [Reddit, it's good enough for production](https://www.reddit.com/r/RedditEng/comments/v1upr8/ipv6_support_on_android/).

**Comment by vaibhav-sharechat** [comment]
When will this be moved to production from alpha ?

**Comment by sauravgk** [comment]
Is happy eyeball also supported for web-socket initial connection?

**Comment by swankjesse** [comment]
Yes.

**Comment by treysis** [comment]
@MichaelJokAr looks like 5.x is still in alpha :(
